core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java

```diff
@@ -212,12 +202,18 @@ static final class HttpResponseWrapper implements StreamWriter<HttpObject> {
         private boolean done;
 
         HttpResponseWrapper(DecodedHttpResponse delegate, @Nullable ClientRequestContext ctx,
-                            RequestLogBuilder logBuilder, HttpRequest request,
                             long responseTimeoutMillis, long maxContentLength) {
             this.delegate = delegate;
             this.ctx = ctx;
-            this.logBuilder = logBuilder;
-            this.request = request;
+            if (ctx != null) {
+                logBuilder = ctx.logBuilder();
+                final HttpRequest request = ctx.request();
+                assert request != null;
+                this.request = request;
+            } else {
+                logBuilder = NOOP_REQUEST_LOG_BUILDER;
+                request = NOOP_HTTP_REQUEST;
```

Comment:
I was rather asking to remove `Noop*` types and just use `null` here. Otherwise, JIT will still see two types are in use in `HttpResponseWrapper`, which is not great for optimization.

Reply:
Updated to create a context for the upgrade request as well. Thanks!